File: solutions/python/dnd-character/1/dnd_character.py
from itertools import islice
from random import randint


def d6() -> int:
    return randint(1, 6)


def random_ability() -> int:
    return sum(islice(sorted((d6() for _ in range(4)), reverse=True), 3))


class Character:
    strength: int
    dexterity: int
    constitution: int
    intelligence: int
    wisdom: int
    charisma: int

    hitpoints: int

    def __init__(self) -> None:
        self.strength = random_ability()
        self.dexterity = random_ability()
        self.constitution = random_ability()
        self.intelligence = random_ability()
        self.wisdom = random_ability()
        self.charisma = random_ability()
        self.hitpoints = 10 + modifier(self.constitution)

    @staticmethod
    def ability() -> int:
        return random_ability()


def modifier(value: int) -> int:
    return (value - 10) // 2

File: solutions/go/dnd-character/1/dnd_character.go
package dndcharacter

import (
	"math/rand"
	"sort"
)

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	return score/2 - 5
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	d6 := func() int {
		return rand.Intn(6) + 1
	}
	fourD6 := []int{d6(), d6(), d6(), d6()}
	sort.Ints(fourD6)
	return fourD6[1] + fourD6[2] + fourD6[3]
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	Constitution := Ability()
	return Character{
		Strength:     Ability(),
		Dexterity:    Ability(),
		Constitution: Constitution,
		Intelligence: Ability(),
		Wisdom:       Ability(),
		Charisma:     Ability(),
		Hitpoints:    10 + Modifier(Constitution),
	}
}
